README.md
# clojure-problems

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-09-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-09-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure-problems/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-problems/compare/0.1.0...0.1.1


test/clojure_problems/core_test.clj
(ns clojure-problems.core-test
  (:require [clojure.test :refer :all]
            [clojure-problems.core :refer :all]))

;; Problem 19
(deftest last-element-test
  (is (= 5 (last-element [1 2 3 4 5])))
  (is (= 3 (last-element '(5 4 3))))
  (is (= "d" (last-element ["b" "c" "d"])))
  (is (= nil (last-element []))))

;; Problem 20
(deftest penultimate-element-test
  (is (= 4 (penultimate-element (list 1 2 3 4 5))))
  (is (= "b" (penultimate-element ["a" "b" "c"])))
  (is (= [1 2] (penultimate-element [[1 2] [3 4]])))
  (is (= nil (penultimate-element []))))

;; Problem 21
(deftest nth-element-test
  (is (= 6 (nth-element '(4 5 6 7) 2)))
  (is (= :a (nth-element [:a :b :c] 0)))
  (is (= 2 (nth-element [1 2 3 4] 1)))
  (is (= [5 6] (nth-element '([1 2] [3 4] [5 6]) 2))))

;; Problem 22
(deftest my-count-test
  (is (= 5 (my-count '(1 2 3 3 1))))
  (is (= 11 (my-count "Hello World")))
  (is (= 3 (my-count [[1 2] [3 4] [5 6]])))
  (is (= 1 (my-count '(13))))
  (is (= 3 (my-count '(:a :b :c)))))

;; Problem 46
(deftest flipping-out-test
  (is (= 3 ((flipping-out nth) 2 [1 2 3 4 5])))
  (is (= true ((flipping-out >) 7 8)))
  (is (= 4 ((flipping-out quot) 2 8)))
  (is (= [1 2 3] ((flipping-out take) [1 2 3 4 5] 3))))

;; Problem 54
(deftest my-partition-test
  (is (= (my-partition 3 (range 9)) '((0 1 2) (3 4 5) (6 7 8))))
  (is (= (my-partition 2 (range 8)) '((0 1) (2 3) (4 5) (6 7))))
  (is (= (my-partition 3 (range 8)) '((0 1 2) (3 4 5)))))

;; Problem 55
(deftest my-frequencies-test
  (is (= (my-frequencies [1 1 2 3 2 1 1]) {1 4, 2 2, 3 1}))
  (is (= (my-frequencies [:b :a :b :a :b]) {:a 2, :b 3}))
  (is (= (my-frequencies '([1 2] [1 3] [1 3])) {[1 2] 1, [1 3] 2})))

;; Problem 56
(deftest my-distinct-test
  (is (= (my-distinct [1 2 1 3 1 2 4]) [1 2 3 4]))
  (is (= (my-distinct [:a :a :b :b :c :c]) [:a :b :c]))
  (is (= (my-distinct '([2 4] [1 2] [1 3] [1 3])) '([2 4] [1 2] [1 3])))
  (is (= (my-distinct (range 50)) (range 50))))

;; Problem 58
(deftest my-comp-test
  (is (= [3 2 1] ((my-comp rest reverse) [1 2 3 4])))
  (is (= 5 ((my-comp (partial + 3) second) [1 2 3 4])))
  (is (= true ((my-comp zero? #(mod % 8) +) 3 5 7 9)))
  (is (= "HELLO" ((my-comp #(.toUpperCase %) #(apply str %) take) 5 "hello world"))))

;; Problem 59
(deftest my-juxt-test
  (is (= [21 6 1] ((my-juxt + max min) 2 3 5 1 6 4)))
  (is (= ["HELLO" 5] ((my-juxt #(.toUpperCase %) count) "hello")))
  (is (= [2 6 4] ((my-juxt :a :c :b) {:a 2, :b 4, :c 6, :d 8 :e 10}))))

;; Problem 60
(deftest my-reductions-test
  (is (= [0 1 3 6 10] (take 5 (my-reductions + (range)))))
  (is (= [[1] [1 2] [1 2 3] [1 2 3 4]] (my-reductions conj [1] [2 3 4])))
  (is (= (reduce * 2 [3 4 5]) 120) (last (my-reductions * 2 [3 4 5]))))

;; Problem 61
(deftest make-map-test
  (is (= (make-map [:a :b :c] [1 2 3]) {:a 1, :b 2, :c 3}))
  (is (= (make-map [1 2 3 4] ["one" "two" "three"]) {1 "one", 2 "two", 3 "three"}))
  (is (= (make-map [:foo :bar] ["foo" "bar" "baz"]) {:foo "foo", :bar "bar"})))

;; Problem 62
(deftest my-iterate-test
  (is (= (take 5 (my-iterate #(* 2 %) 1)) [1 2 4 8 16]))
  (is (= (take 100 (my-iterate inc 0)) (take 100 (range))))
  (is (= (take 9 (my-iterate #(inc (mod % 3)) 1)) (take 9 (cycle [1 2 3])))))

;; Problem 63
(deftest my-group-by-test
  (is (= (my-group-by #(> % 5) [1 3 6 8]) {false [1 3], true [6 8]}))
  (is (= (my-group-by #(apply / %) [[1 2] [2 4] [4 6] [3 6]])
         {1/2 [[1 2] [2 4] [3 6]], 2/3 [[4 6]]}))
  (is (= (my-group-by count [[1] [1 2] [3] [1 2 3] [2 3]])
         {1 [[1] [3]], 2 [[1 2] [2 3]], 3 [[1 2 3]]})))

;; Problem 69
(deftest my-merge-with-test
  (is (= (my-merge-with * {:a 2, :b 3, :c 4} {:a 2} {:b 2} {:c 5})
         {:a 4, :b 6, :c 20}))
  (is (= (my-merge-with - {1 10, 2 20} {1 3, 2 10, 3 15})
         {1 7, 2 10, 3 15}))
  (is (= (my-merge-with concat {:a [3], :b [6]} {:a [4 5], :c [8 9]} {:b [7]})
         {:a [3 4 5], :b [6 7], :c [8 9]})))

;; Problem 132
(deftest insert-between-test
  (is (= '(1 :less 6 :less 7 4 3) (insert-between < :less [1 6 7 4 3])))
  (is (= '(2) (insert-between > :more [2])))
  (is (= [0 1 :x 2 :x 3 :x 4]  (insert-between #(and (pos? %) (< % %2)) :x (range 5))))
  (is (empty? (insert-between > :more ())))
  (is (= [0 1 :same 1 2 3 :same 5 8 13 :same 21]
         (take 12 (->> [0 1]
                       (iterate (fn [[a b]] [b (+ a b)]))
                       (map first)             ; fibonacci numbers
                       (insert-between (fn [a b] ; both even or both odd
                                         (= (mod a 2) (mod b 2)))
                                       :same))))))

;; Problem 158
(deftest decurry-test
  (is (= 10 ((decurry (fn [a]
                        (fn [b]
                          (fn [c]
                            (fn [d]
                              (+ a b c d))))))
             1 2 3 4)))
  (is (= 24 ((decurry (fn [a]
                        (fn [b]
                          (fn [c]
                            (fn [d]
                              (* a b c d))))))
             1 2 3 4)))
  (is (= 25 ((decurry (fn [a]
                        (fn [b]
                          (* a b))))
             5 5))))


src/clojure_problems/core.clj
(ns clojure-problems.core)

(defn last-element
  "
  Problem 19: Write a function which returns the last element in a sequence.
  http://www.4clojure.com/problem/19
  "
  [[x & xs]]
  (if xs (last-element xs) x))

(defn penultimate-element
  "
  Problem 20: Write a function which returns the second to last element
  from a sequence.
  http://www.4clojure.com/problem/20
  "
  [[x y & xs]]
  (if xs (recur (cons y xs)) x))

(defn nth-element
  "
  Problem 21: Write a function which returns the Nth element from a sequence.
  http://www.4clojure.com/problem/21
  "
  [xs n]
  ((zipmap (range (count xs)) xs) n))

(defn my-count
  "
  Problem 22: Write a function which returns the total number of elements
  in a sequence.
  http://www.4clojure.com/problem/22
  "
  [xs]
  (reduce (fn [acc _] (inc acc)) 0 xs))

(defn rotate-seq [n xs])

(defn flipping-out
  "
  Problem 46: Write a higher-order function which flips the order of the
  arguments of an input function.
  http://www.4clojure.com/problem/46
  "
  [f]
  (fn [x y] (f y x)))

(defn my-partition
  "
  Problem 54: Write a function which returns a sequence of lists of x items
  each. Lists of less than x items should not be returned.
  http://www.4clojure.com/problem/54
  "
  [n ls]
  (when (<= n (count ls))
    (cons  (take n ls) (my-partition n (drop n ls)))))

(defn my-frequencies
  "
  Problem 55: Write a function which returns a map containing the number of
  occurences of each distinct item in a sequence.
  http://www.4clojure.com/problem/55
  "
  [coll]
  (reduce (fn [acc v]
            (if (acc v)
              (update-in acc [v] inc)
              (assoc acc v 1)))
          {}
          coll))

(defn my-distinct
  "
  Problem 56: Write a function which removes the duplicates from a sequence.
  Order of the items must be maintained.
  http://www.4clojure.com/problem/56
  "
  [coll]
  (:ret (reduce (fn [acc v]
                  (if ((:visited acc) v)
                    acc
                    (-> acc
                        ;; Why does 4Clojure not know `update`..?
                        (update-in [:ret] conj v)
                        (update-in [:visited] conj v))))
                {:ret [] :visited #{}}
                coll)))

(defn my-comp
  "
  Problem 58: Write a function which allows you to create function compositions.
  The parameter list should take a variable number of functions, and create a
  function that applies them from right-to-left.
  http://www.4clojure.com/problem/58
  "
  [& fns]
  (fn [& args]
    (reduce (fn [acc f] (f acc))
            (apply (first (reverse fns)) args)
            (rest (reverse fns)))))

(defn my-juxt
  "
  Problem 59: Take a set of functions and return a new function that takes a
  variable number of arguments and returns a sequence containing the result of
  applying each function left-to-right to the argument list.
  http://www.4clojure.com/problem/59
  "
  [& fns]
  (fn [& args]
    (map #(apply % args) fns)))

(defn my-reductions
  "
  Problem 60: Write a function which behaves like reduce, but returns each
  intermediate value of the reduction. Your function must accept either two or
  three arguments, and the return sequence must be lazy.
  "
  ([f coll]
   (my-reductions f (first coll) (rest coll)))
  ([f prev coll]
   (if (seq (rest coll))
     (lazy-seq (cons prev (my-reductions f (f prev (first coll)) (rest coll))))
     (list prev (f prev (first coll))))))

(defn make-map
  "
  Problem 61: Write a function which takes a vector of keys and a vector of
  values and constructs a map from them.
  http://www.4clojure.com/problem/61
  "
  [ks vs]
  (apply merge (map #(assoc {} %1 %2) ks vs)))

(defn my-iterate
  "
  Problem 62: Given a side-effect free function f and an initial value x write a
  function which returns an infinite lazy sequence of x, (f x), (f (f x)), (f (f
  (f x))), etc.
  http://www.4clojure.com/problem/62
  "
  [f x]
  (lazy-seq (cons x (my-iterate f (f x)))))

(defn my-group-by
  "
  Problem 63: Given a function f and a sequence s, write a function which
  returns a map. The keys should be the values of f applied to each item in s.
  The value at each key should be a vector of corresponding items in the order
  they appear in s.
  http://www.4clojure.com/problem/63
  "
  [f coll]
  (reduce (fn [acc v]
            (let [k (f v)]
              (if (acc k)
                ;; Weird, 4Clojure doesn't recognize `update`
                ;; but does recognize `update-in`..
                (update-in acc [k] concat [v])
                (assoc acc k [v]))))
          {}
          coll))

(defn my-merge-with
  "
  Problem 69: Write a function which takes a function f and a variable number of
  maps. Your function should return a map that consists of the rest of the maps
  conj-ed onto the first. If a key occurs in more than one map, the mapping(s)
  from the latter (left-to-right) should be combined with the mapping in the
  result by calling (f val-in-result val-in-latter).
  http://www.4clojure.com/problem/69
  "
  [f m0 & ms]
  (reduce (fn [acc m]
            (merge acc
                   (reduce (fn [acc1 [k v]]
                             (if (acc k)
                               (assoc acc1 k (f (acc k) v))
                               (assoc acc1 k v)))
                           {}
                           m)))
          m0
          ms))

(defn insert-between
  "
  Problem 132: Write a function that takes a two-argument predicate, a value,
  and a collection; and returns a new collection where the value is inserted
  between every two items that satisfy the predicate.
  http://www.4clojure.com/problem/132
  "
  [pred v coll]
  (->> coll
       (partition 2 1 nil)
       (map (fn [[x y]]
              (cond
                (nil? y) (list x)
                (pred x y)  (list x v)
                :else (list x))))
       (apply concat)))

(defn decurry
  "
  Problem 158: Write a function that accepts a curried function of unknown arity
  n. Return an equivalent function of n arguments.
  http://www.4clojure.com/problem/158
  "
  [f]
  (fn [& args]
    (reduce (fn [acc v] (acc v))
            (f (first args))
            (rest args))))
